### PR TITLE
add etcd and kube-apiserver faults

### PR DESCRIPTION
**Pull request by xiaojingchen**
this pr contains follow changes:
- add some checkers
  - CheckK8sAvailable
  - CheckOperatorAvailable
  - CheckTidbClustersAvailable
- add etcd/apiserver stop faults

### CODE REVIEW

File: tests/cmd/stability/main.go [hunk 1]
```diff
@@ -214,5 +220,89 @@ func main() {
 		oa.CheckTidbClusterStatusOrDie(cluster)
 	}
 
+	// stop one etcd node and k8s/operator/tidbcluster is available
+	faultEtcd := selectNode(conf.ETCDs)
+	err := fta.StopETCD(faultEtcd)
```
Comment (weekface) [inline tests/cmd/stability/main.go]: extract it as `StopAEtcdOrDie`

File: tests/cmd/stability/main.go [hunk 2]
```diff
@@ -214,5 +220,89 @@ func main() {
...
+	if err != nil {
+		glog.Fatal(err)
+	}
+	defer fta.StartETCD(faultEtcd)
+	err = tests.Keep(3*time.Second, 10*time.Minute, func() error {
```
Comment (weekface) [inline tests/cmd/stability/main.go]: ditto

File: tests/util.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+import (
+	"time"
+
+	"github.com/pingcap/tidb-operator/pkg/client/clientset/versioned"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+func CreateKubeClient() (versioned.Interface, kubernetes.Interface, error) {
```
Comment (weekface) [inline tests/util.go]: use `cli, kubeCli := client.NewCliOrDie()`

File: tests/cmd/stability/main.go [hunk 4]
```diff
@@ -40,6 +42,7 @@ func main() {
 	oa := tests.NewOperatorActions(cli, kubeCli, conf)
 	fta := tests.NewFaultTriggerAction(cli, kubeCli, conf)
 	fta.CheckAndRecoverEnvOrDie()
+	oa.CheckK8sAvailable(nil, nil)
```
Comment (weekface) [inline tests/cmd/stability/main.go]: This function has a return error

File: tests/cmd/stability/main.go [hunk 5]
```diff
@@ -219,5 +222,19 @@ func main() {
 	// truncate a sst file and check failover
 	oa.TruncateSSTFileThenCheckFailoverOrDie(cluster1, 5*time.Minute)
 
+	// stop one etcd node and k8s/operator/tidbcluster is available
+	faultEtcd := tests.SelectNode(conf.ETCDs)
```
Comment (weekface) [inline tests/cmd/stability/main.go]: this `SelectNode` method will select the first etcd forever: https://github.com/pingcap/tidb-operator/blob/7b9ced3b34d57e9a9d8d1fb82e8915ced652fa87/tests/manifests/stability/stability-configmap.yaml#L19

File: tests/failover.go [hunk 6]
```diff
@@ -445,3 +446,238 @@ func (oa *operatorActions) GetNodeMap(info *TidbClusterConfig, component string)
...
+		if err != nil {
+			return err
+		}
+		glog.Infof("all clusters is available")
+		return nil
+	})
+}
+
+func (oa *operatorActions) CheckK8sAvailable(excludeNodes map[string]*corev1.Node, excludePods map[string]*corev1.Pod) error {
+	return wait.Poll(3*time.Second, time.Minute, func() (bool, error) {
```
Comment (weekface) [inline tests/failover.go]: use the default interval

File: tests/failover.go [hunk 7]
```diff
@@ -445,3 +446,238 @@ func (oa *operatorActions) GetNodeMap(info *TidbClusterConfig, component string)
...
+				glog.Errorf("pod:[%s/%s] is unavailable,state is %s", pod.GetName(), pod.GetNamespace(), podState)
+				return false, nil
+			}
+		}
+		return true, nil
+	})
+}
+
+func (oa *operatorActions) CheckOperatorAvailable(operatorConfig *OperatorConfig) error {
+	return wait.Poll(3*time.Second, 3*time.Minute, func() (bool, error) {
```
Comment (weekface) [inline tests/failover.go]: use the default interval

File: tests/failover.go [hunk 8]
```diff
@@ -445,3 +446,238 @@ func (oa *operatorActions) GetNodeMap(info *TidbClusterConfig, component string)
...
+		}
+		if schedulerDeployment.Status.AvailableReplicas != *schedulerDeployment.Spec.Replicas {
+			return false, fmt.Errorf("the %s is not available", tidbSchedulerName)
+		}
+		return true, nil
+	})
+}
+
+func (oa *operatorActions) CheckTidbClustersAvailable(infos []*TidbClusterConfig) error {
+	return wait.Poll(3*time.Second, 30*time.Second, func() (bool, error) {
```
Comment (weekface) [inline tests/failover.go]: use the default interval

File: tests/failover.go [hunk 9]
```diff
@@ -445,3 +446,238 @@ func (oa *operatorActions) GetNodeMap(info *TidbClusterConfig, component string)
...
+	_, err = db.Exec(fmt.Sprintf("INSERT INTO %s VALUES (?)", testTableName), "testValue")
+	if err != nil {
+		glog.Infof("cluster:[%s] can't insert data to mysql: %v", info.FullName(), err)
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func GetPodStatus(pod *corev1.Pod) string {
```
Comment (weekface) [inline tests/failover.go]: Why use such a complicated function?  just `.status.phase` is not ok?

File: tests/cmd/stability/main.go [hunk 10]
```diff
@@ -219,5 +222,19 @@ func main() {
...
+	fta.StopETCDOrDie(faultEtcd)
+	defer fta.StartETCDOrDie(faultEtcd)
+	oa.CheckOneEtcdDownOrDie(operatorCfg, allClusters, faultEtcd)
```
Comment (weekface) [inline tests/cmd/stability/main.go]: You should add other cases: stopping 2 etcds and stopping 3 etcds.

File: tests/cmd/stability/main.go [hunk 11]
```diff
@@ -219,5 +222,19 @@ func main() {
 	// truncate a sst file and check failover
 	oa.TruncateSSTFileThenCheckFailoverOrDie(cluster1, 5*time.Minute)
 
+	// stop one etcd node and k8s/operator/tidbcluster is available
+	faultEtcd := tests.SelectNode(conf.ETCDs)
+	fta.StopETCDOrDie(faultEtcd)
+	defer fta.StartETCDOrDie(faultEtcd)
```
Comment (weekface) [inline tests/cmd/stability/main.go]: before `CheckOneEtcdDownOrDie ` you should sleep about 30 minutes.

File: tests/failover.go (hunk 6 above)
Comment (xiaojingchen) [reply tests/failover.go]: default interval is too long for the case

File: tests/failover.go (hunk 9 above)
Comment (xiaojingchen) [reply tests/failover.go]: yes, the `.status.phase` just is the pod phase, but not pod's real state

File: tests/cmd/stability/main.go (hunk 10 above)
Comment (xiaojingchen) [reply tests/cmd/stability/main.go]: Sure, I will add these cases in next pr

File: tests/failover.go (hunk 9 above)
Comment (weekface) [reply tests/failover.go]: Where do these codes come from?

**Comment by xiaojingchen** [comment]
@weekface PTAL

**Pull request merged**